File: cpp/consensus_string.cpp


#include <iostream>
#include <limits>
#include <list>
#include <map>
#include <utility>

using namespace std;

void copy_bases(std::list<string> &seqs, std::map<char, int> bases) {
    for(std::map<char, int>::iterator it=bases.begin(); it != bases.end(); it++)
        seqs.push_back(string(1, it->first));
}

void print_bases(std::map<char, int> bases) {
    std::cout << "bases to be appened" << std::endl;
    for(std::map<char, int>::iterator it = bases.begin(); it != bases.end(); it++)
        if (it->first)
            std::cout << it->first << "\t" << it->second << std::endl;
}


void print_consensus_str(std::list<string> consensus_str, int seq_len) {
    for(std::list<string>::iterator it=consensus_str.begin(); it != consensus_str.end(); it++) {
        std::cout << (*it).length() << "\t" << (*it) << std::endl;
    }
}


void print_seqs(string** seqs, int num_tests, int num_seqs) {
    std::cout<<std::endl;
    std::cout << "Print out seqs from stdin" << std::endl;
    for(int i=0; i < num_tests; i++) {
        for(int j=0; j < num_seqs+1; j++) {
            std::cout << seqs[i][j] << std::endl;
        }
    }
}


std::list<string> get_consensus_seqs(string *seqs, int num_seqs, int seq_len) {
    std::list<string> base_string;
    for(int i=0; i<seq_len; i++) {
        std::map<char, int> base_map;
        int max_count = 0;
        for(int j=0; j<num_seqs+1; j++) {
            char b = seqs[j][i];
            if (base_map.count(b) == 0)
                base_map[b] = 1;
            else base_map[b] += 1;

            if (base_map[b] > max_count)
                max_count = base_map[b];
        }
        if (base_string.size() == 0) {
            for(std::map<char, int>::iterator it = base_map.begin(); it != base_map.end(); it++) {
                if (it->first && (it->second == max_count)) {
                    base_string.push_back(string(1, it->first));
                }
            }
        } else {
            int q_size = base_string.size();
            while (q_size > 0) {
                string bstr = base_string.front();
                base_string.pop_front();
                for (std::map<char, int>::iterator it=base_map.begin(); it != base_map.end(); it++) {
                    if (it->first) {
                        base_string.push_back(bstr + string(1, it->first));
                    }
                }
                q_size--;  
            }
        }
    }
    return base_string;
}

bool compare_lex(std::pair<string, int> p1, std::pair<string, int> p2) {
    if (p1.first < p2.first)
        return true;
    else return false;
}


int hamming_distance(string seq1, string seq2) {
    int h_dist = 0;
    for (int i=0; i < seq1.size(); i++)
        if (seq1[i] != seq2[i])
            h_dist++;
    return h_dist;
}

std::list<std::pair<string, int> > find_min_consensus(std::list<string> seqs, std::list<int> errors) {
    int min_error = std::numeric_limits<int>::max();
    std::list<std::pair<string, int> > consensus_index;
    std::list<string>::iterator sit = seqs.begin();
    for (std::list<int>::iterator it = errors.begin(); it != errors.end(); it++) {
        if ((*it) <= min_error) {
            min_error = *it;
        }
        sit++;
    }

    sit = seqs.begin();
    for (std::list<int>::iterator it = errors.begin(); it != errors.end(); it++) {
        if ((*it) == min_error) {
            consensus_index.push_back(std::make_pair(*sit, *it));
        }
        sit++;
    }
    return consensus_index;
}

std::pair<string, int> find_consensus_error(std::list<string> &consensus_str, string *seqs, int num_seqs) {
    std::list<int> consensus_error;
    for (std::list<string>::iterator it=consensus_str.begin(); it != consensus_str.end(); it++) {
        int con_error = 0;
        for (int i=0; i < num_seqs+1; i++) {
            if (seqs[i].size() > 0) {
                int dist = hamming_distance(*it, seqs[i]);
                con_error += dist;
            }
        }
        consensus_error.push_back(con_error); 
    }
    std::list<std::pair<string, int> > seq_info = find_min_consensus(consensus_str, consensus_error);
    if (seq_info.size() > 1)
        seq_info.sort(compare_lex);
    return seq_info.front();
}

int main() {
    int num_tests, num_seqs, seq_len;
    int n = 0;
    std::cout <<"num_tests: ";
    std::cin >> num_tests;
    string** test_seqs = new string *[num_tests];
    int test_num_seqs[num_tests+1];
    int test_seq_len[num_tests+1];
    while (n < num_tests) {
        std::cout << "num_seqs seq_len: ";
        std::cin >> num_seqs >> seq_len;
        test_num_seqs[n] = num_seqs;
        test_seq_len[n] = seq_len;
        test_seqs[n] = new string[num_seqs+1];
        int m = 0;
        while (m  < num_seqs+1) {
            std::getline(std::cin, test_seqs[n][m]);
            m++;
        }
        n++;
    }
    n = 0;
    while (n < num_tests) {
        std::list<string> consensus_str = get_consensus_seqs(test_seqs[n], test_num_seqs[n], test_seq_len[n]);
        std::pair<string, int> seq_info = find_consensus_error(consensus_str, test_seqs[n], test_num_seqs[n]);
        n++;

        std::cout << "Consensus seq: " << seq_info.first << std::endl;
        std::cout << "Consensus error: " << seq_info.second << std::endl;
    }
return 0;

}

File: python/consensus_string.py

from collections import deque
import sys
import time


def hamm_distance(chaine1, chaine2):
    return sum(c1 != c2 for c1, c2 in zip(chaine1, chaine2))

def hamming_distance(seq1, seq2):
    if len(seq1) != len(seq1):
        raise ValueError("Sequence length should be same.")
    h_dist = 0
    for s1, s2 in zip(seq1, seq2):
        if s1 != s2:
            h_dist +=1
    return h_dist


def find_min_consensus(con_errors):
    min_error = sys.maxsize
    consensus_index = []
    for i, e in enumerate(con_errors):
        if e <= min_error:
            min_error = e
            consensus_index.append(i)
    return consensus_index
    

def count_bases(bases):
    freq = {}
    for b in bases:
        if b not in freq:
            freq[b] = 0
        freq[b] += 1
    return freq.keys()


def get_max_base(freq):
    max_freq = -1
    max_bases = []
    for k, v in freq.items():
        if v > max_freq:
            max_freq = v
    for k, v in freq.items():
        if v == max_freq:
            max_bases.append(k)
    return max_bases


def get_consensus_string(seqs):
    base_string = None
    for i in range(len(seqs[0])):
        bases = [seq[i] for seq in seqs]
        max_bases = count_bases(bases)
        #max_bases = get_max_base(freq)
        if not base_string:
            base_string = deque(max_bases)
        else:
            dq_len = len(base_string)
            while dq_len > 0:
                bstr = base_string.popleft()
                for b in max_bases:
                    base_string.append(bstr+b)
                dq_len -= 1
    return base_string


def find_consensus_error(consensus_seqs, seqs):
    consensus_error = []
    for con_seq in consensus_seqs:
        con_error = 0
        for i in range(len(seqs)):
            dist = hamm_distance(con_seq, seqs[i])
            con_error += dist
        consensus_error.append(con_error)
    
    index = find_min_consensus(consensus_error)
    if len(index) == 1:
        return (consensus_seqs[0], consensus_error[index[0]])
    else:
        seq_info = list([(consensus_seqs[i], consensus_error[i]) for i in index])
        seq_info.sort(key=lambda x: x[0])
        return seq_info[0]
                

def main(argv):
    start_time = time.time()
    print("start_time: {}".format(start_time))
    num_tests = int(input("Num tests: "))
    n = num_tests
    test_seqs = []
    while n > 0:
        seq_info = input("num_seq seq_len: ")
        seq_info = seq_info.split(' ')
        assert len(seq_info) == 2
        num_seqs, seq_len = int(seq_info[0]), int(seq_info[1])
        seqs = [input() for i in range(num_seqs)]
        for seq in seqs:
            assert seq_len == len(seq)
        test_seqs.append(seqs)
        print("Number of tests: {}".format(num_tests))
        print("Number of sequences: {}".format(num_seqs))
        print("Sequence length: {}".format(seq_len))
        print("seqs: {}".format(seqs))
        n -= 1
    n = 0
    while n < num_tests:
        consensus_str = get_consensus_string(test_seqs[n])
        seq_info = find_consensus_error(consensus_str, test_seqs[n])
        print("Consensus seq: {}".format(seq_info[0]))
        print("Consensus error: {}".format(seq_info[1]))
        n += 1
    end = time.time()
    print("Time being taken:L {}".format(time.time() - start_time))

if __name__ == "__main__":
    main(sys.argv[1:])
